README.md
TurnServer
==========
A fork from [http://turnserver.sourceforge.net](http://turnserver.sourceforge.net/index.php?n=Main.HomePage)

TurnServer is an open-source implementation of Traversal Using Relays around NAT
(TURN) protocol. It aims to be compliant with RFC5766 (TURN) and RFC5389 (STUN).

The TURN protocol allows a client to obtain IP addresses and ports from such a
relay. It is most useful for elements behind symmetric NATs or firewalls that
wish to be on the receiving end of a connection to a single peer.

TURN clients can connect to TurnServer with the following protocols: UDP, TCP
and TLS over TCP. Experimental DTLS support is also provided. Relaying data can
be done with UDP or TCP protocol.

TurnServer supports also RFC5389 (STUN Binding request), RFC6062 (relay data
with TCP protocol) and RFC6156 (relay IPv6-IPv6, IPv4-IPv6 and IPv6-IPv4).

TurnServer is known to work on the following systems:
- GNU/Linux 2.6;
- FreeBSD 7.x, 8.x.

1) Build / install
------------------

TurnServer requires following libraries:
- libconfuse development files (version >= 2.6);
- libssl development files;
- librt (normally included in Linux and *BSD distribution*).

TurnServer is written in pure C according to the C99 and POSIX + XSI standards.
Thus it should be compiled on all POSIX systems which have realtime signals
support.

Note for *BSD* users, install the required libconfuse ports in /usr/ prefix,
otherwise you have to set the PKG_CONFIG_PATH variable or make symlinks before
running ./configure script:

    ln -sf /usr/local/lib/libconfuse.so /usr/lib/ && \
    ln -sf /usr/local/include/confuse.h /usr/include/

To build TurnServer, run following commands:

    $ autoreconf -i
    $ ./configure
    $ make
    $ make install

./configure can take options:

    --enable-debug-build                 : allow to compile with debug informations
                                           default=no
    --enable-fdsetsize=number            : allow to preconfigure FD_SETSIZE macro
                                           (must be a number >=32) default=no
    --enable-xor-peer-address-max=number : allow to preconfigure
                                           XOR_PEER_ADDRESS_MAX macro (must be a
                                           number > 0) default=5

Copy the template configuration file (extra/turnserver.conf.template) and
template accounts database file (extra/turnusers.txt) to a directory of your
choice (i.e. /etc/ or /usr/local/etc/).
Do not forget, the accounts database file pathname has to be populated in
configuration file (attribute account_file). See next sections to know how to
setup configuration and accounts files.

To generate the API documentation:

    $ make doxygen-run

The HTML generated documentation is located in doc/html/ directory of TurnServer sources.

Launch the server:

    $ turnserver -c /path/to/config/file

2) Configuration file
---------------------

In extra/ directory you will find a configuration template file
(turnserver.conf.template). Change settings according to your environment.

Here are important parameters,
- listen_address        : public IPv4 address;
- listen_addressv6      : public IPv6 address;
- realm                 : realm (i.e. domain.org) of the server;
- account_file          : specify the location of the accounts database file;
- tcp_port and udp_port : bind the service on the specified port;
- tls                   : enable TLS support;
- tls_port              : bind the secure service on the specified port.
- ca_file               : Certification Authority (must set if tls = true)
- cert_file             : server certificate (must set if tls = true)
- private_key_file      : server private key (must set if tls = true)
- turn_tcp              : enable TURN-TCP extension
- tcp_buffer_userspace  : enable userspace buffering for TURN-TCP extension, if
                          false OS buffering will be used
- tcp_buffer_size       : maximum amount of bytes that can be buffered for
                          TURN-TCP (RFC6062) extension

Other parameters such as allocations number quota or experimental features are
documented in manpages:

    $ man turnserver.conf

3) Accounts database file
--------------------------

TurnServer uses (for the moment) a basic text file which contains accounts
information.

The format of each line is:

    login:password:realm:state

The state can be "authorized", "refused" or "restricted". The "restricted" state
means the account has bandwidth restrictions.

Note: realm have to match realm parameter defined in TurnServer configuration
file. The ":" character is also forbidden in login, password or realm fields.

4) Security
------------

If TurnServer is launched as root or set-uid root, it is possible to drop
privileges.

One possibility is to create a special user (which have less privileges). To
create such a user named turnserver:

    adduser --system --group turnserver

Then you have to tell configuration file to choose this user:

    unpriv_user = turnserver

If TurnServer is set-uid root and unpriv_user is not set, TurnServer will drop
privileges to the user who launched the binary.

**Note**: if turnserver is launched as root and unpriv_user not set, the program
will not loose its root privileges.

5) How-to test simply turnserver
--------------------------------

TurnServer is shipped with two test tools: test_turn_client and 
test_echo_server. The first one is a minimal TURN client and test_echo_server
is a simple UDP echo server.

To test TurnServer simply:
- configure turnserver.conf;
- configure turnusers.txt ;
- launch "turnserver -c /path/to/turnserver.conf";
- launch "test_echo_server 8086";
- launch "test_turn_client -t udp -s turnserver_address -p turnserver_address -w 8086 -u user -g password -d domain.org".

The turnserver_address parameter should be the address configured in
turnserver.conf's listen_address or listen_addressv6. if you want to use
localhost here, you should configure listen_address to 127.0.0.1 _and_ 
listen_addressv6 to ::1. The user, password and domain.org parameters are the
ones from turnusers.txt.

It is not necessary to run the server and the test tools on different computers
but it is recommended just to be sure everything work as in real use-case.



ChangeLog
2013-06-14	Sebastien Vincent <sebastien.vincent@turnserver.org>

	* Version 0.7.3		 - Minor fix to be able to compile with recent glibc.

2012-11-19	Daniel Pocock <daniel@pocock.com.au>

  * Version 0.7.2		 - Further tweaks to autotools

2012-10-17	Sebastien Vincent <sebastien.vincent@turnserver.org>

	* Version 0.7.1		 - Minor modifications to ease Debian package integration.

2012-10-01  Sebastien Vincent <sebastien.vincent@turnserver.org>

  * Version 0.7      - Fixes compilation warnings with clang;
                     - Send indications do not need to have realm, user, ...;
                     - Fixes typos.

2012-02-22  Sebastien Vincent <sebastien.vincent@turnserver.org>

  * Version 0.6      - Add spec file to build an RPM package (by Ken Cox);
                     - Better syslog logging separated by level (by Ken Cox):
                     - Bug fixes provided by Ryo Sato and Ken Cox;
                     - Fixes TURN-TCP bugs reported by Nigel Pattinson.

2011-06-10  Sebastien Vincent <sebastien.vincent@turnserver.org>

  * Version 0.5      - Add restricted users feature;
                     - Add RFC6062 (TURN-TCP) document;
                     - Add RFC6156 (TURN-IPV6) document;
                     - Add mod_tmpuser to create/delete temporary users;
                     - Fix endianess.

2010-05-02  Sebastien Vincent <sebastien.vincent@turnserver.org>

  * Version 0.4     - Add RFC5766 (TURN) document;
                    - Add turn-ipv6-09 and turn-tcp-06 drafts support;
                    - Add multiprotocol client example (UDP/TCP/TLS/DTLS);
                    - Add allocation port number range configuration;
                    - Some files can compiled with MS Windows as well as
                      multiprotocol client example.

2010-01-03  Sebastien Vincent <sebastien.vincent@turnserver.org>

  * Version 0.3     - Add turn-ipv6-08 draft support;
                    - Add credentials reload support;
                    - Fix a "stack smashing";
                    - Experimental DTLS support (not in TURN standard).

2009-08-26  Sebastien Vincent <sebastien.vincent@turnserver.org>

  * Version 0.2.3   - Add turn-ipv6-06 draft support;
                    - Add turn-16 draft support;
                    - Fix lifetime bug.

2009-03-09  Sebastien Vincent <vincent@lsiit.u-strasbg.fr>

  * Version 0.2.2   - Load sharing with multiple uplink addresses;
                    - Increase nonce lifetime to one hour;
                    - Fix ChannelBind refresh;
                    - Add turn-13 draft;
                    - Add 403 error message.

2008-12-17  Sebastien Vincent <vincent@lsiit.u-strasbg.fr>

  * Version 0.2.1   - Fix CRC32 generation;
                    - Drop privileges if run as root/setuid root.

2008-12-15  Sebastien Vincent <vincent@lsiit.u-strasbg.fr>

  * Version 0.2     - Add attributes from turn-12 draft;
                    - Add authentication on all TURN responses when possible;
                    - Pre-calculate account MD5 hash;
                    - Re-organize some code (message-integrity calculation, ...);
                    - Alternate behavior for DF flag;
                    - Separate TCP and TLS sockets;
                    - Correctly relayed data from peer to client in TLS;
                    - Multiple XOR-PEER-ADDRESS support in CreatePermission;
                    - Update wireshark parser (packet-stun2.c);
                    - Additionnal configuration parameters;
                    - Allocation number and bandwidth quota;
                    - Address/port restrictions;
                    - Reconstruct fragmented TCP messages support;
                    - Prevent allocation hijacking support;
                    - Add logs support via syslog.

2008-11-22  Sebastien Vincent <vincent@lsiit.u-strasbg.fr>

  * Version 0.1.3   - Backport code refactoring from 0.2-draft-11 branch;
                    - Additionnal configuration parameters.

2008-10-21  Sebastien Vincent <vincent@lsiit.u-strasbg.fr>

  * Version 0.1.2.2 - Fix nonce validation for 64-bit platform.

2008-09-24  Sebastien Vincent <vincent@lsiit.u-strasbg.fr>

  * Version 0.1.2.1 - Fix bug in go_daemon().

2008-09-02  Sebastien Vincent <vincent@lsiit.u-strasbg.fr>

  * Version 0.1.2   - Fix bug (infinite loop) when receiving unknown 
                      comprehension-required attribute.

2008-08-16  Sebastien Vincent <vincent@lsiit.u-strasbg.fr>

  * Version 0.1.1   - Fix IPv6 problem;
                    - Add missing XOR for FINGERPRINT mechanism;
                    - Fix minors bugs.

2008-07-26  Sebastien Vincent <vincent@lsiit.u-strasbg.fr>

  * Initial version.



src/conf.c
/*
 *  TurnServer - TURN server implementation.
 *  Copyright (C) 2008-2009 Sebastien Vincent <sebastien.vincent@turnserver.org>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  In addition, as a special exception, the copyright holders give
 *  permission to link the code of portions of this program with the
 *  OpenSSL library under certain conditions as described in each
 *  individual source file, and distribute linked combinations
 *  including the two.
 *  You must obey the GNU General Public License in all respects
 *  for all of the code used other than OpenSSL.  If you modify
 *  file(s) with this exception, you may extend this exception to your
 *  version of the file(s), but you are not obligated to do so.  If you
 *  do not wish to do so, delete this exception statement from your
 *  version.  If you delete this exception statement from all source
 *  files in the program, then also delete it here.
 */

/**
 * \file conf.c
 * \brief Configuration parsing.
 * \author Sebastien Vincent
 * \date 2008-2009
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdlib.h>
#include <string.h>

#include <sys/types.h>
#include <sys/socket.h>

#include <arpa/inet.h>

#include <confuse.h>

#include "conf.h"
#include "turnserver.h"

/**
 * \brief Free the resources used by the lex parser.
 *
 * This function comes from libconfuse and is not called
 * in cfg_free(), that's why call it here.
 * \note Require libconfuse >= 2.6.
 * \return 0
 */
extern int cfg_yylex_destroy(void);

/**
 * \var g_denied_address_opts
 * \brief Denied address option.
 */
static cfg_opt_t g_denied_address_opts[] =
{
  CFG_STR("address", "", CFGF_NONE),
  CFG_INT("mask", 24, CFGF_NONE),
  CFG_INT("port", 0, CFGF_NONE),
  CFG_END()
};

/**
 * \var g_opts
 * \brief Options recognized.
 */
static cfg_opt_t g_opts[]=
{
  CFG_STR("listen_address", NULL, CFGF_LIST),
  CFG_STR("listen_addressv6", NULL, CFGF_LIST),
  CFG_INT("udp_port", 3478, CFGF_NONE),
  CFG_INT("tcp_port", 3478, CFGF_NONE),
  CFG_INT("tls_port", 5349, CFGF_NONE),
  CFG_BOOL("tls", cfg_false, CFGF_NONE),
  CFG_BOOL("dtls", cfg_false, CFGF_NONE),
  CFG_INT("max_port", 65535, CFGF_NONE),
  CFG_INT("min_port", 49152, CFGF_NONE),
  CFG_BOOL("turn_tcp", cfg_false, CFGF_NONE),
  CFG_BOOL("tcp_buffer_userspace", cfg_true, CFGF_NONE),
  CFG_INT("tcp_buffer_size", 1500, CFGF_NONE),
  CFG_INT("restricted_bandwidth", 10, CFGF_NONE),
  CFG_BOOL("daemon", cfg_false, CFGF_NONE),
  CFG_STR("unpriv_user", NULL, CFGF_NONE),
  CFG_INT("max_client", 50, CFGF_NONE),
  CFG_INT("max_relay_per_username", 10, CFGF_NONE),
  CFG_INT("allocation_lifetime", 1800, CFGF_NONE),
  CFG_STR("nonce_key", NULL, CFGF_NONE),
  CFG_STR("ca_file", NULL, CFGF_NONE),
  CFG_STR("cert_file", NULL, CFGF_NONE),
  CFG_STR("private_key_file", NULL, CFGF_NONE),
  CFG_STR("realm", "domain.org", CFGF_NONE),
  CFG_STR("account_method", "file", CFGF_NONE),
  CFG_STR("account_file", "users.txt", CFGF_NONE),
  CFG_SEC("denied_address", g_denied_address_opts, CFGF_MULTI),
  CFG_INT("bandwidth_per_allocation", 0, CFGF_NONE),
  CFG_BOOL("mod_tmpuser", cfg_false, CFGF_NONE),
  /* the following attributes are not used for the moment */
  CFG_STR("account_db_login", "anonymous", CFGF_NONE),
  CFG_STR("account_db_password", "anonymous", CFGF_NONE),
  CFG_STR("account_db_name", "turnserver", CFGF_NONE),
  CFG_STR("account_db_address", "127.0.0.1", CFGF_NONE),
  CFG_INT("account_db_port", 3306, CFGF_NONE),
  CFG_END()
};

/**
 * \var g_cfg
 * \brief Config pointer.
 */
static cfg_t* g_cfg = NULL;

int turnserver_cfg_parse(const char* file, struct list_head* denied_address_list)
{
  int ret = 0;
  size_t i = 0;
  size_t nb = 0;
  g_cfg = cfg_init(g_opts, CFGF_NONE);

  ret = cfg_parse(g_cfg, file);

  if(ret == CFG_FILE_ERROR)
  {
    fprintf(stderr, "Cannot find configuration file %s\n", file);
    return -1;
  }
  else if(ret == CFG_PARSE_ERROR)
  {
    fprintf(stderr, "Parse error in configuration file %s\n", file);
    return -2;
  }

  /* check IPv4 listen addresses to be valid IPv4 ones */
  nb = cfg_size(g_cfg, "listen_address");
  for(i = 0 ; i < nb ; i++)
  {
    struct sockaddr_storage addr;
    char* str = cfg_getnstr(g_cfg, "listen_address", i);
    if(inet_pton(AF_INET, str, &addr) != 1)
    {
      return -2;
    }
  }

  /* check IPv6 listen addresses to be valid IPv6 ones */
  nb = cfg_size(g_cfg, "listen_addressv6");
  for(i = 0 ; i < nb ; i++)
  {
    struct sockaddr_storage addr;
    char* str = cfg_getnstr(g_cfg, "listen_addressv6", i);
    if(inet_pton(AF_INET6, str, &addr) != 1)
    {
      return -2;
    }
  }

  /* add the denied address */
  nb = cfg_size(g_cfg, "denied_address");
  for(i = 0 ; i < nb ; i++)
  {
    cfg_t* ad = cfg_getnsec(g_cfg, "denied_address", i);
    char* addr = cfg_getstr(ad, "address");
    uint8_t mask = cfg_getint(ad, "mask");
    uint16_t port = cfg_getint(ad, "port");
    struct denied_address* denied = NULL;

    if(!(denied = malloc(sizeof(struct denied_address))))
    {
      return -3;
    }

    memset(denied, 0x00, sizeof(struct denied_address));
    denied->mask = mask;
    denied->port = port;

    if(inet_pton(AF_INET, addr, denied->addr) != 1)
    {
      /* try IPv6 */
      if(inet_pton(AF_INET6, addr, denied->addr) != 1)
      {
        free(denied);
        return -2;
      }
      else
      {
        /* check mask */
        if(mask > 128)
        {
          free(denied);
          return -2;
        }
        denied->family = AF_INET6;
      }
    }
    else
    {
      /* mask check */
      if(mask > 24)
      {
        free(denied);
        return -2;
      }
      denied->family = AF_INET;
    }

    /* add to the list */
    LIST_ADD(&denied->list, denied_address_list);
  }

  return 0;
}

void turnserver_cfg_print(void)
{
  fprintf(stdin, "Configuration:\n");
  cfg_print(g_cfg, stderr);
}

void turnserver_cfg_free(void)
{
  if(g_cfg)
  {
    cfg_free(g_cfg);
    g_cfg = NULL;

    cfg_yylex_destroy();
  }
}

char* turnserver_cfg_listen_address(void)
{
  size_t nb = cfg_size(g_cfg, "listen_address");

  if(nb)
  {
    size_t l = (size_t)(rand() % nb);
    return cfg_getnstr(g_cfg, "listen_address", l);
  }
  else
  {
    return NULL;
  }
}

char* turnserver_cfg_listen_addressv6(void)
{
  size_t nb = cfg_size(g_cfg, "listen_addressv6");

  if(nb)
  {
    size_t l = (size_t)(rand() % nb);
    return cfg_getnstr(g_cfg, "listen_addressv6", l);
  }
  else
  {
    return NULL;
  }
}

uint16_t turnserver_cfg_udp_port(void)
{
  return cfg_getint(g_cfg, "udp_port");
}

uint16_t turnserver_cfg_tcp_port(void)
{
  return cfg_getint(g_cfg, "tcp_port");
}

uint16_t turnserver_cfg_tls_port(void)
{
  return cfg_getint(g_cfg, "tls_port");
}

int turnserver_cfg_tls(void)
{
  return cfg_getbool(g_cfg, "tls");
}

int turnserver_cfg_dtls(void)
{
  return cfg_getbool(g_cfg, "dtls");
}

uint16_t turnserver_cfg_max_port(void)
{
  return cfg_getint(g_cfg, "max_port");
}

uint16_t turnserver_cfg_min_port(void)
{
  return cfg_getint(g_cfg, "min_port");
}

int turnserver_cfg_turn_tcp(void)
{
  return cfg_getbool(g_cfg, "turn_tcp");
}

int turnserver_cfg_tcp_buffer_userspace(void)
{
  return cfg_getbool(g_cfg, "tcp_buffer_userspace");
}

uint32_t turnserver_cfg_tcp_buffer_size(void)
{
  return cfg_getint(g_cfg, "tcp_buffer_size");
}

uint32_t turnserver_cfg_restricted_bandwidth(void)
{
  return cfg_getint(g_cfg, "restricted_bandwidth");
}

int turnserver_cfg_daemon(void)
{
  return cfg_getbool(g_cfg, "daemon");
}

char* turnserver_cfg_unpriv_user(void)
{
  return cfg_getstr(g_cfg, "unpriv_user");
}

uint16_t turnserver_cfg_max_client(void)
{
  return cfg_getint(g_cfg, "max_client");
}

uint16_t turnserver_cfg_max_relay_per_username(void)
{
  return cfg_getint(g_cfg, "max_relay_per_username");
}

uint16_t turnserver_cfg_allocation_lifetime(void)
{
  return cfg_getint(g_cfg, "allocation_lifetime");
}

char* turnserver_cfg_nonce_key(void)
{
  return cfg_getstr(g_cfg, "nonce_key");
}

char* turnserver_cfg_ca_file(void)
{
  return cfg_getstr(g_cfg, "ca_file");
}

char* turnserver_cfg_cert_file(void)
{
  return cfg_getstr(g_cfg, "cert_file");
}

char* turnserver_cfg_private_key_file(void)
{
  return cfg_getstr(g_cfg, "private_key_file");
}

char* turnserver_cfg_realm(void)
{
  return cfg_getstr(g_cfg, "realm");
}

uint16_t turnserver_cfg_bandwidth_per_allocation(void)
{
  return cfg_getint(g_cfg, "bandwidth_per_allocation");
}

char* turnserver_cfg_account_method(void)
{
  return cfg_getstr(g_cfg, "account_method");
}

char* turnserver_cfg_account_file(void)
{
  return cfg_getstr(g_cfg, "account_file");
}

char* turnserver_cfg_account_db_login(void)
{
  return cfg_getstr(g_cfg, "account_db_login");
}

char* turnserver_cfg_account_db_password(void)
{
  return cfg_getstr(g_cfg, "account_db_password");
}

char* turnserver_cfg_account_db_name(void)
{
  return cfg_getstr(g_cfg, "account_db_name");
}

char* turnserver_cfg_account_db_address(void)
{
  return cfg_getstr(g_cfg, "account_db_address");
}

uint16_t turnserver_cfg_account_db_port(void)
{
  return cfg_getint(g_cfg, "account_db_port");
}

int turnserver_cfg_mod_tmpuser(void)
{
  return cfg_getbool(g_cfg, "mod_tmpuser");
}
